### PR TITLE
MAINT:stats: Type annotations for _sobol.pyx

### LINKED ISSUE #14038: MAINT: add type annotations for _sobol.pyx
Following what have been done for `scipy.stats._qmc_cy.pyx`, we could add easily types to `scipy.stats._sobol.pyx`. Types for Cython cannot be inline but have to be added using a `.pyi` file.

### PR DESCRIPTION
**Pull request by akpandeya**
#### Reference issue
Closes #14038

#### What does this implement/fix?
Types added to scipy.stats._sobol.pyx

### CODE REVIEW

**Review by tupui — CHANGES_REQUESTED** [review]
It's a good start @akpandeya, thanks!

In order to test the changes, you need to remove mypy skipping from `mypy.ini`:

```ini
# Remove these 2 lines
[mypy-scipy.stats._sobol]
ignore_missing_imports = True
```

And I guess that all the functions will then need to be annotated.

As for the title of the issue, could you rename it to make it clearer? No need to put the referring issue here as it's in the description (plus it is confusing as we have 2 numbers following each other like that). Also, remember that all commit should use our naming convention using the keyword and a comprehensive description.

**Comment by ilayn** [comment]
Thanks @akpandeya, looks good. Just a passing by comment. 

Just for the future reference, please make separate branches to work on new features and to send PRs such that your standard working repo does not interfere with the PRs you have submitted. Once a PR is merged you can safely delete that branch and keep working on other branches. 

You can leave the `closes #` etc ... comments to the Github issue description here since Github will remember what the issue was. But sometimes if it is in the commit message  then the issue starts to get pinged from various forks. Not a big deal but just a detail.

**Comment by akpandeya** [comment]
> Thanks @akpandeya, looks good. Just a passing by comment.
> 
> Just for the future reference, please make separate branches to work on new features and to send PRs such that your standard working repo does not interfere with the PRs you have submitted. Once a PR is merged you can safely delete that branch and keep working on other branches.
> 
> You can leave the `closes #` etc ... comments to the Github issue description here since Github will remember what the issue was. But sometimes if it is in the commit message then the issue starts to get pinged from various forks. Not a big deal but just a detail.

Thank you very much for all the suggestions. I will keep them in mind.

File: scipy/stats/_sobol.pyi
Comment (BvB93) [inline scipy/stats/_sobol.pyi:4]:
```suggestion
import numpy.typing as npt
```
FYI, `if TYPE_CHECKING:` checks are redundant within .pyi stub files, 
as the expression _always_ evaluates to `True` in such cases.

Comment (tupui) [reply scipy/stats/_sobol.pyi:4]: oh thanks, we should then remove this from the other `pyi` files. This is something you could do here @akpandeya (optional of course).

Comment (BvB93) [reply scipy/stats/_sobol.pyi:4]: > Side note: sorry we did not have time to go through your point! I added it to the next topic list.

Yeah, it was a _big_ list though, so not a huge suprise.

Comment (akpandeya) [reply scipy/stats/_sobol.pyi:4]: ok.  Thanks for the comment. Will improve the import.

**Comment by tupui** [comment]
Hey @akpandeya just checking in. Can I help you with something?

**Comment by akpandeya** [comment]
> Hey @akpandeya just checking in. Can I help you with something?

Hi, Thanks for checking in. 4 tests are failing locally for me.

**Comment by tupui** [comment]
> Hi, Thanks for checking in. 4 tests are failing locally for me.

It's fine these are not related to your changes so you can update the PR. In this case, the only test which will be affected will be the type annotation one (mypy). So if you want to check this locally first, you just need to run this one.

**Comment by tupui** [comment]
You can see in the [logs](https://github.com/scipy/scipy/pull/14061/checks?check_run_id=2637816881) that the CI failed for your changes now.

```bash
scipy/stats/_qmc.py:1034: error: Argument 2 to "initialize_v" has incompatible type "Union[int, integer[Any]]"; expected "int"  [arg-type]
scipy/stats/_qmc.py:1056: error: Argument 1 to "_cscramble" has incompatible type "Union[int, integer[Any]]"; expected "int"  [arg-type]
scipy/stats/_qmc.py:1084: error: Argument 3 to "_draw" has incompatible type "Union[int, integer[Any]]"; expected "int"  [arg-type]
scipy/stats/_qmc.py:1088: error: Argument 1 to "_draw" has incompatible type "Union[int, integer[Any]]"; expected "int"  [arg-type]
scipy/stats/_qmc.py:1088: error: Argument 3 to "_draw" has incompatible type "Union[int, integer[Any]]"; expected "int"  [arg-type]
scipy/stats/_qmc.py:1152: error: Argument 3 to "_fast_forward" has incompatible type "Union[int, integer[Any]]"; expected "int"  [arg-type]
scipy/stats/_qmc.py:1155: error: Argument 1 to "_fast_forward" has incompatible type "Union[int, integer[Any]]"; expected "int"  [arg-type]
scipy/stats/_qmc.py:1155: error: Argument 3 to "_fast_forward" has incompatible type "Union[int, integer[Any]]"; expected "int"  [arg-type]
```

**Comment by tupui** [comment]
Some pointers: for the first error, you need to replace `Int` with `IntNumber` in `initialize_v `. You have to do this for the other functions which are reported to have the issue.

Also this needs to be added at the top to have `IntNumber`.
```
from scipy._lib._util import IntNumber
```

**Comment by akpandeya** [comment]
> Some pointers: for the first error, you need to replace `Int` with `IntNumber` in `initialize_v `. You have to do this for the other functions which are reported to have the issue.
> 
> Also this needs to be added at the top to have `IntNumber`.
> 
> ```
> from scipy._lib._util import IntNumber
> ```
Thanks for the pointers. I will update soon.

**Comment by akpandeya** [comment]
> Some pointers: for the first error, you need to replace `Int` with `IntNumber` in `initialize_v `. You have to do this for the other functions which are reported to have the issue.
> 
> Also this needs to be added at the top to have `IntNumber`.
> 
> ```
> from scipy._lib._util import IntNumber
> ```

There is no IntNumber in scipy._lib._util, though

**Comment by tupui** [comment]
> > Some pointers: for the first error, you need to replace `Int` with `IntNumber` in `initialize_v `. You have to do this for the other functions which are reported to have the issue.
> > Also this needs to be added at the top to have `IntNumber`.
> > ```
> > from scipy._lib._util import IntNumber
> > ```
> 
> There is no `IntNumber` in `scipy._lib._util`, though

You have to merge master to your branch.

**Review by tupui — APPROVED** [review]
LGTM, thanks @akpandeya for your first contribution 😃

@BvB93 do you have any other comment before I merge?

**Comment by akpandeya** [comment]
> LGTM, thanks @akpandeya for your first contribution 😃
> 
> @BvB93 do you have any other comment before I merge?

It would not have been possible without all the help I got from you and others in this thread. I learnt a lot l. Hopefully, will be able contribute more

**Comment by ilayn** [comment]
When everyone's happy with thecomments, one of the members from the team will click the merge button so you don't need to close it, especially not after all the hardwork you put in 😉

**Comment by tupui** [comment]
Thanks @ilayn for reopening. This is basically ready. I am just waiting for a confirmation because type hints being tricky, there might be some fancy things I missed.

File: scipy/stats/_sobol.pyi
Comment (BvB93) [inline scipy/stats/_sobol.pyi:46]:
```suggestion
from typing_extensions import Literal  # Preferrably place this import at the top of the file

_MAXDIM: Literal[21201]
_MAXBIT: Literal[30]
_MAXDEG: Literal[18]
```
There are two proposed changes here:
* First off, `_MAXDEG` currently seems to be missing from the stubs. https://github.com/scipy/scipy/blob/4ec6f29c0344ddce80845df0c2b8765fc8f27a72/scipy/stats/_sobol.pyx#L28-L30
* Secondly,  these three integers all have a constant pre-defined value, so we can use [literal](https://docs.python.org/3/library/typing.html#typing.Literal) integers here to provide just a bit more information.

File: scipy/stats/_sobol.pyi
Comment (BvB93) [inline scipy/stats/_sobol.pyi:5]:
```suggestion
import numpy as np
from scipy._lib._util import IntNumber

def initialize_v(
    v : np.ndarray, 
```
@tupui I'm not sure I'm interpreting this correctly, but does this signature not mean that the function operates on an ndarray, and not arbitrary array-likes?
https://github.com/scipy/scipy/blob/4ec6f29c0344ddce80845df0c2b8765fc8f27a72/scipy/stats/_sobol.pyx#L188
If so, it seems that all instances of `npt.ArrayLike` will have to be replaced with `np.ndarray` throughout the file.

Comment (tupui) [reply scipy/stats/_sobol.pyi:5]: Indeed, but then the current types in `_qmc_cy.pyi` are also not correct (this is even clearer there as the functions are calling things like `sample_view.shape[0]`). I guess it went through the crack for simplicity.

@akpandeya would you like to make the changes? It would be great if you could do it for `_qmc_cy.pyi` as well then. Let me know, I can also do this before merging.

File: scipy/stats/_sobol.pyi
Comment (tupui) [reply scipy/stats/_sobol.pyi:46]: Ok for the literals, makes sense. But then I guess we should also do it in `_qmc.py` as these where defined as `ClassVar[int]`?

Comment (BvB93) [reply scipy/stats/_sobol.pyi:46]: That would definitely be an improvement, yes, though it is not necessarily mandatory. 
Namely, `ClassVar[int]` is still a super-type of `ClassVar[Literal[30]]`, so this change shouldn't introduce any new mypy errors.

File: scipy/stats/_sobol.pyi
Comment (akpandeya) [reply scipy/stats/_sobol.pyi:5]: Sure thing! Will do it by tomorrow.

**Comment by tylerjereddy** [comment]
@tupui @BvB93 ready to merge?

File: scipy/stats/_sobol.pyi
Comment (tylerjereddy) [reply scipy/stats/_sobol.pyi:5]: I guess this needs to be addressed? If you want this PR to get merged for the 1.7.0 release try to get the change done soon, I will branch quite soon, targeting Wednesday

**Review by BvB93 — APPROVED** [review]
LGTM with the latest changes, though let's wait and see if the CI brings something up.

**Review by tupui — APPROVED** [review]
LGTM as well, thanks @akpandeya and @BvB93 for the review. I will wait for the CI and merge afterwards.

**Comment by tupui** [comment]
@tylerjereddy shall I add this to your PR about the release notes, or do you generate the list again before the release?

**Comment by tylerjereddy** [comment]
I'll re-run the script to generate the list yeah

**Pull request merged**
